internal/webhook/core/v1/pod_webhook.go

```diff
@@ -0,0 +1,188 @@
+/*
+Copyright 2024.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/go-logr/logr"
+	"github.com/opendatahub-io/odh-model-controller/internal/controller/constants"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// +kubebuilder:webhook:path=/mutate--v1-pod,mutating=true,failurePolicy=fail,groups="",resources=pods,verbs=create,versions=v1,name=mutating.pod.odh-model-controller.opendatahub.io,reinvocationPolicy=IfNeeded,admissionReviewVersions=v1,sideEffects=none
+
+var podlog logr.Logger
+
+// Mutator is a webhook that injects incoming pods
+type PodMutatorDefaultor struct {
+	Client client.Client
+}
+
+var _ webhook.CustomDefaulter = &PodMutatorDefaultor{}
+
+// Handle incoming Pod and executes mutation logic.
+func (m *PodMutatorDefaultor) podMutator(pod *corev1.Pod) error {
+	if !needMutate(pod) {
+		return nil
+	}
+
+	if err := m.mutate(pod); err != nil {
+		podlog.Error(err, "Failed to mutate pod", "name", pod.Name, "namespace", pod.Namespace)
+		return err
+	}
+
+	return nil
+}
+
+func (m *PodMutatorDefaultor) mutate(pod *corev1.Pod) error {
+	if needToAddRayTLSGenerator(pod) {
+		rayTLSGeneratorScript := getRayTLSGeneratorScriptInitContainer()
+		pod.Spec.InitContainers = append(pod.Spec.InitContainers, *rayTLSGeneratorScript)
+	}
+	return nil
+}
+
+// Add ray tls generator init-container into the pod when the pod has pod environment variable "RAY_USE_TLS" is 1.
+func getRayTLSGeneratorScriptInitContainer() *corev1.Container {
+
+	script := `
+echo "Generating Self Signed Certificate for Ray nodes"
+REFORMAT_IP=$(echo ${POD_IP} | sed 's+\.+\\.+g')
+JSONPATH={.data.$REFORMAT_IP}
+RAY_PEM_CONTENT=""
+RAY_CA_PEM_CONTENT=$(oc get secret ray-tls -n $POD_NAMESPACE -o jsonpath="{.data.ca\.crt}")
```

Comment:
If you think a start-up delay of 1-2 minutes is tolerable, I suggest you to mount the TLS secret to the pod.

Secrets are synced to the pod when they change (with some delay). This script can be limited to check and wait for the certs to become available by simply reading the mounted Secret files. You also have the advantage that you no longer need any Roles nor RoleBindings.

Reply:
My original code follows exactly the approach you described. However, after running actual tests, I found that 1–2 minutes is much longer than expected. If I were the user, I would think this is too slow.
So, although it’s a bit more complex, I decided to use the CLI to handle this part as quickly as possible.

@danielezonca @terrytangyuan @spolti WDYT?

Reply:
> I found that 1–2 minutes is much longer than expected

You mean that 1-2 minutes feels too much? or in your tests it consistently was longer than 1-2 minutes?

If it was the former, I think adding a 1-2 minute wait to the already long time for downloading a big model isn't that much. If it is the later, I can understand this approach.

However, when weighting the security implications of the role (my [other comment](https://github.com/opendatahub-io/odh-model-controller/pull/346#discussion_r1935079945)), I will be more comfortable with a longer wait time, in case there is no easy way to reduce attack surface.

Reply:
it was the both cases. downloading will be happening after generating cert. So it always plus 1~2 mins to deploy model. Using current way, I see it is getting faster. The first deploy take more time then second time. but original one is always  slow. it is slower than configmap update as well.  
I tested the time ConfigMap/Secret but I end up to change code. 

Moreover, testing is much worse. It will extend testing time a lot. Regarding the role, I will think about it tomorrow.

Reply:
updated role part.